compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -407,11 +408,39 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
 
                 err.span_label(span, cannot_infer_msg);
 
-                let printer = fmt_printer(self, Namespace::TypeNS);
-                let args = printer.comma_sep(generic_args.iter().copied()).unwrap().into_buffer();
+                let args: Result<_, std::fmt::Error> = try {
+                    let mut printer = fmt_printer(self, Namespace::TypeNS);
+                    let print = |arg: GenericArg<'tcx>, mut printer: FmtPrinter<'a, 'tcx>| {
+                        if arg.is_suggestable(self.tcx) {
+                            printer = arg.print(printer)?;
+                        } else {
+                            printer.write_str(match arg.unpack() {
+                                GenericArgKind::Lifetime(_) => "'_",
+                                GenericArgKind::Type(_) | GenericArgKind::Const(_) => "_",
+                            })?;
+                        }
+                        Ok(printer)
+                    };
+
+                    let mut generic_args = generic_args.iter().copied();
+                    if let Some(arg) = generic_args.next() {
+                        printer = print(arg, printer)?;
+                        for arg in generic_args {
+                            printer.write_str(", ")?;
+                            printer = print(arg, printer)?;
+                        }
+                    }
+
+                    printer.into_buffer()
+                };
+                let args = args.unwrap();
+
                 err.span_suggestion_verbose(
                     insert_span,
-                    &format!("consider specifying the generic argument{}", pluralize!(args.len()),),
+                    &format!(
+                        "consider specifying the generic argument{}",
+                        pluralize!(generic_args.len()),
```

Comment:
`args.len()` was incorrect: `args` is a `String` here, not a vector of arguments. It was measuring the amount of bytes in the string, not the number of arguments.